File: frontend/src/types/Transfer.ts
import { ShipmentStatus } from "@/enums/shipmentStatus";
import { TaxInclusionType } from "@/enums/TaxInclusionType";

export interface Transfer {
  id: number;
  fromWarehouse: number;
  toWarehouse: number;
  date: string;
  orderTax: string;
  discount: string;
  shippingCost: string;
  grandTotal: string;
  status: ShipmentStatus;
  note?: string;
  products: TransferProduct[];
  createdBy?: number;
  createdAt?: string;
  updatedBy?: number;
  updatedAt?: string | null;
  companyId?: number;
}

export interface TransferProduct {
  productId: number;
  productCode: string;
  productUnitCost: string;
  productStock: number;
  transferredQty: number;
  productDiscount: string;
  productTax: string;
  subTotal: string;
}

export interface SelectedTransferProduct {
  productId: number;
  productName: string;
  code: string;
  stock: number;
  cost: string;
  discount: string;
  tax: string;
  inclusionType: TaxInclusionType;
  subTotal: string;
  transferredQty: number;
}

export interface CreateTransferRequest {
  fromWarehouse: number;
  toWarehouse: number;
  date: string;
  orderTax: string;
  discount: string;
  shippingCost: string;
  status: ShipmentStatus;
  note?: string;
  products: {
    productId: number;
    productCode: string;
    productUnitCost: string;
    productStock: number;
    transferredQty: number;
    productDiscount: string;
    productTax: string;
    subTotal: string;
  }[];
}

File: backend/src/main/java/nextpos/app/nextpos/model/entity/Transfer.java
package nextpos.app.nextpos.model.entity;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import nextpos.app.nextpos.model.enums.ShipmentStatus;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "transfer")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Transfer {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "from_warehouse_id", nullable = false)
    private Warehouse fromWarehouse;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "to_warehouse_id", nullable = false)
    private Warehouse toWarehouse;

    @Column(nullable = false)
    private LocalDate date;

    @Builder.Default
    @OneToMany(mappedBy = "transfer", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<TransferProduct> products = new ArrayList<>();

    @Column(columnDefinition = "TEXT")
    private String note;

    @Column(name = "order_tax", precision = 10, scale = 2)
    private BigDecimal orderTax;

    @Column(name = "discount", precision = 10, scale = 2)
    private BigDecimal discount;

    @Column(name = "shipping_cost", precision = 10, scale = 2)
    private BigDecimal shippingCost;

    @Column(name = "grand_total", precision = 15, scale = 2)
    private BigDecimal grandTotal;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 50)
    private ShipmentStatus status;

    @Column(name = "created_by")
    private Long createdBy;

    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_by")
    private Long updatedBy;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "company_id")
    private Long companyId;
}
